### PR TITLE
Use a thread pool backed sequential executor for DSL rules and events

### PR DESCRIPTION
**Pull request by joerg1985**
This is basically a reincarnation of the idea in #2182 with a different approach.

The `SequentialScheduledExecutorService` in this PR will act like a `Executors.newSingleThreadScheduledExecutor`, but it does use a thread pool to execute the tasks. The mechanism to block the `ScheduledExecutorService` to run tasks concurrently is based on a chain of `CompletableFutures`. Each instance has a reference to the last CompletableFuture and will call `handleAsync` to add a new task. This will ensure tasks are running sequentially. 
A `.schedule` call will register a callback which will then submit the task to perform now, ending in the same handleAsync call on the `CompletableFutures`, to ensure these are also running sequentially.

The only side effect a rule has, is to not run inside the same thread again and again. This might break things like a `ThreadLocal`, therefore a Rule has to opt-in to this new behavior (see `Rule.isThreadBound`).

This feature is currently disabled by default and has to be enabled by setting the system property  `org.openhab.core.common.SequentialScheduledExecutorService` to `true`.
 
When deserializing an existing `RuleTemplate` the `threadBound` flag should be set true. I am not sure how to do this, i tried in the `RuleTemplateDTO` by setting `threadBound` true for a new instance.

Further work: The rule DSLRuleProvider could inspect the DSL and set `threadBound` to true when a `ThreadLocal` or `Thread` is used by the rule. This will ensure no exisiting DSL rule will break.

### CODE REVIEW

**Comment by joerg1985** [comment]
It might be better to use VirtualThreads as soon as java 21 is the minimum version, this would make things easier. Less code and ThreadLocals would still work without special handling.

**Comment by kaikreuzer** [comment]
It will still take quite some time before we move to Java 21 as a minimal requirement...

**Comment by joerg1985** [comment]
@kaikreuzer okay, so i will push this further. Do you think it does make sense to have this `Rule.isThreadBound` logic or is this over engineered? The jruby rule support requesting it in the old tickets is deprecated and if no one else might need this?

**Comment by kaikreuzer** [comment]
@joerg1985 I was asking myself where we would need it. For the DSL rules I don't think it is necessary; besides scheduling timers, there is no thread support available and in the past we already had thread pools for the execution in place.
So unless anyone from the scripting side (@rkoshak?) asks for it, I would leave this feature aside.

**Comment by rkoshak** [comment]
I'm not sure I understand what the "it" is here under discussion. 

From my perspective, in the old days there was a relatively limited thread pool for rules and any individual rule could be running multiple times across more than one thread. This could and often did lead to all the rules failing to run because of just one long running rule ran amok. I want to avoid reintroducing that behavior.

It's also my understanding that before the introduction of the `cache`, the only reason users of Nashorn JS scripts in the UI was a side effect that the rule's context was reused for each run of the rule: `this.foo = (this.foo === undefined) ? 'initial value' : this.foo;`. This line means if `this.foo` already exists, use that, otherwise initialize it to `initial value`. 

I don't know if that worked because the thread was reused or something else. I don't think we necessarily need to be overly concerned with how Nashorn UI rules worked but it would break some user's rules if this behavior were to change.

Finally, there is an issue where the first time a JS Scripting (and therefore Blockly) script is loaded it can take up to 30 seconds to run the first time on 32-bit ARM machines (i.e. the recommended openHABian config). If reusing threads also means reloading rules this problem will become really bad as it won't just be the first time the rule runs, it will be random or it will be every time.

So, as long as we retain the situation where one rule running amok can't impact other rule's ability to run, I'm happy. As long as using a thread pool doesn't require reloading a script for each thread it runs in, I'm happy.  I do not want to go back to the days where we needed to implement reentrant locks in our rules to prevent them from running more than once at a time and with or without that a poorly written rule can bring down *all* rules by exhausting the thread pool.

**Comment by rkoshak** [comment]
> Is there any need to ensure some rules must be called on the same thread all the time?

I don't know enough about the implications. 

Does running the rule on another thread require the Script Actions/Conditions to be reloaded or reinitialized? 
  - If not, this should be transparent to the users and it's all good
  - If so, the change may break some older Nashorn JS Script Actions/Conditions and it will cause problems for JS Scripting and Blockly users. 

There is a problem in JS Scripting right now where it can take 20-40 seconds to run the rule that first time after it's loaded/reloaded. Having to randomly reload the scripts if it's put on a different Thread would be a nightmare.

**Comment by joerg1985** [comment]
@kaikreuzer Okay it looks like there might be issues, but as long as the user has to opt-in via the `org.openhab.core.common.SequentialScheduledExecutorService` system property everything should be fine.
In some years when the JDK 21 is used, the system property can be removed and the `ThreadPoolManager.newSequentialScheduledExecutorService` could return a `Executors.newSingleThreadExecutor` backed by a virtual thread by default.

So i would suggest to:
- move the `org.openhab.core.common.SequentialScheduledExecutorService` check inside the `ThreadPoolManager.newSequentialScheduledExecutorService`
- update the arguments of `ThreadPoolManager.newSequentialScheduledExecutorService` to switch to the virtual threads without changing the signature of the method.
- remove the `threadBound` flag from the Rule

Any concerns about this?

**Comment by kaikreuzer** [comment]
If I get it right, with the new system property, it would be a "hidden" feature that brave users can try out by activating it. For all others there won't be any change, correct?
If so, yes: With this approach we could get some experience with the behavior and see whether it has any negative impact on the initialization of JS rules. If this is not the case, we might one day decide to make this the default and only allow opt-outs.

**Comment by joerg1985** [comment]
@kaikreuzer you are right, this wil protect from trouble.

**Comment by joerg1985** [comment]
@kaikreuzer i just updated the PR, the flag is gone and the review should be easier now.

**Comment by joerg1985** [comment]
@kaikreuzer is there any thing left that i need to do, before this can be merged?
Just asking, because i have some time the next days.

**Comment by morph166955** [comment]
So while I don't disagree with the concept, please please please please please make sure that there is a way to turn this off for power users.  I've had so many issues with thread pools causing congestion over the last few years that the concept of adding one is terrifying to me.  I have a ton of rules that run concurrently.  I also understand that I'm a power user and as such my OH is running on a relatively beefy device that can handle such a load.  Again, I can absolutely understand the reason for this especially on lower end platforms, it's why I wrote the original PR that you noted above, just please make sure there is a way to run this without a pool as well.  Thank you!

**Comment by joerg1985** [comment]
@morph166955 the current plan is to opt-in via a system property to avoid general issues. 

Regarding the congestion of the pool, this PR uses a unlimited pool, in the worst case the number of threads in the pool is equal to the number of rules. There will be some latency to run new threads, but this are only milliseconds.

This sounds like you could be the perfect tester for this, or at least share some numbers? The total number of rules and the number if concurrent executing rules would be interesting for me.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/internal/events/EventHandler.java [hunk 1]
```diff
@@ -69,9 +70,8 @@ public EventHandler(final Map<String, Set<EventSubscriber>> typedEventSubscriber
     }
 
     private synchronized ExecutorRecord createExecutorRecord(Class<? extends EventSubscriber> subscriber) {
-        return new ExecutorRecord(
-                Executors.newSingleThreadExecutor(new NamedThreadFactory("eventexecutor-" + executors.size())),
-                new AtomicInteger());
+        return new ExecutorRecord(ThreadPoolManager.newSequentialScheduledExecutorService("events-pool",
```
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/internal/events/EventHandler.java:73]:
```suggestion
        return new ExecutorRecord(ThreadPoolManager.newSequentialScheduledExecutorService("events",
```

File: bundles/org.openhab.core/src/test/java/org/openhab/core/common/ThreadPoolManagerTest.java [hunk 2]
```diff
@@ -35,6 +42,135 @@
...
+
+    @Test
+    public void testExecutionIsSequentialInSequentialExecutorService() {
+        ScheduledExecutorService service = ThreadPoolManager.newSequentialScheduledExecutorService("unit-test",
+                "thread-1");
+
+        AtomicBoolean done = new AtomicBoolean(false);
+
+        service.submit(() -> {
+            Thread.sleep(1000);
```
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/test/java/org/openhab/core/common/ThreadPoolManagerTest.java:89]: Could we reduce the waiting to `100` throughout the tests or does that make them instable? I would try to reduce the execution time as much as possible.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+
+            scheduled.add((RunnableFuture<?>) future);
+            origin.complete((RunnableFuture<?>) future);
+
+            return future;
+        }
+    }
+
+    @Override
+    public <V> ScheduledFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit) {
```
Comment (kaikreuzer) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java:122]: Can we somehow avoid the code duplication here (and in the next two methods)?

**Review by kaikreuzer — CHANGES_REQUESTED** [review]
Sorry that it took me so long - please find some review comments below.
I have to admit that I do not fully understand the implemented mechanism here. In general, my experience with implementing own executor services is that this is a very tricky job and that a lot of things can go wrong. So I would definitely consider this feature for now experimental and we should indeed only enable it on an opt-in basis for testing.  Nonetheless, I'd love to see it working and reducing the number of threads that are used, so let's move forward with it!

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+
+    public SequentialScheduledExecutorService(ScheduledThreadPoolExecutor pool) {
+        if (pool.getMaximumPoolSize() != Integer.MAX_VALUE) {
+            throw new IllegalArgumentException("the pool must scale unlimited to avoid potential dead locks!");
+        }
+
+        this.pool = pool;
+
+        // prepare the WorkQueueEntry we are using when no tasks are pending
+        RunnableCompletableFuture<?> future = new RunnableCompletableFuture();
```
[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java:64]: will be fixed in the next push

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+ * {@link CompletableFuture}s.
+ * Each instance has a reference to the last CompletableFuture and will call handleAsync to add a new task.
+ *
+ * @author Jörg Sautter - Initial contribution
+ */
+class SequentialScheduledExecutorService implements ScheduledExecutorService {
+
+    private final WorkQueueEntry empty;
+    private final ScheduledThreadPoolExecutor pool;
+    private final List<RunnableFuture<?>> scheduled;
```
[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java:52]: All accesses to the list are inside a `synchronized (this)` block.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+        future.complete(null);
+        empty = new WorkQueueEntry(null, null, future);
+
+        // tracks scheduled tasks alive
+        this.scheduled = new ArrayList<>();
+
+        tail = empty;
+
+        // clean up to ensure we do not keep references to old tasks
+        cleaner = this.scheduleAtFixedRate(() -> {
```
[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java:74]: will be fixed in the next push

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+ * This is a drop in replacement to a ScheduledExecutorService with one thread to avoid a lot of threads created, idling
+ * most of the time and wasting memory on low-end devices.
+ *
+ * The mechanism to block the ScheduledExecutorService to run tasks concurrently is based on a chain of
+ * {@link CompletableFuture}s.
+ * Each instance has a reference to the last CompletableFuture and will call handleAsync to add a new task.
+ *
+ * @author Jörg Sautter - Initial contribution
+ */
+class SequentialScheduledExecutorService implements ScheduledExecutorService {
```
[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java:48]: okay

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/ThreadPoolManager.java [hunk 8]
```diff
@@ -115,6 +116,26 @@ protected void modified(Map<String, Object> properties) {
...
+     * Returns an instance of a scheduled service, which will sequentially execute submitted tasks. If a task is
+     * currently running the task is queued until the previous one is completed, this also applies for scheduled tasks.
+     * The service might execute submitted task might in different threads, but still one after the other.
+     * If it is the first request for the given pool name and a pool is used, the instance is newly created.
+     *
+     * @param poolName a short name used to identify the pool, if a thread pool is used e.g. "discovery"
+     * @param threadName a short name used to identify the thread if no thread pool is used, e.g. "bluetooth"
+     * @return an instance to use
+     */
+    public static ScheduledExecutorService newSequentialScheduledExecutorService(String poolName, String threadName) {
```
[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/ThreadPoolManager.java:129]: I renamed it to getPoolBasedSequentialScheduledExecutorService

File: bundles/org.openhab.core/src/test/java/org/openhab/core/common/ThreadPoolManagerTest.java (hunk 2 above)
Comment (joerg1985) [reply bundles/org.openhab.core/src/test/java/org/openhab/core/common/ThreadPoolManagerTest.java:89]: 100ms should be more than enought

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java [hunk 9]
```diff
@@ -0,0 +1,537 @@
...
+        Callable<?> callable = () -> {
+            task.run();
+
+            return null;
+        };
+
+        return submit0(origin, callable);
+    }
+
+    private <T> CompletableFuture<T> submit0(RunnableFuture<?> origin, Callable<T> task) {
```
[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java:294]: you are right, e.g. `submitToWorkQueue` is a better name.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java [hunk 10]
```diff
@@ -0,0 +1,537 @@
...
+
+        // wait for any future to complete
+        while (true) {
+            boolean allDone = true;
+
+            for (CompletableFuture<T> future : futures) {
+                if (future.isDone()) {
+                    if (!future.isCompletedExceptionally()) {
+                        // stop the others
+                        for (CompletableFuture<T> toLate : futures) {
```
[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java:397]: yes it is tooLate, will change it.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+        empty = new WorkQueueEntry(null, null, future);
+
+        // tracks scheduled tasks alive
+        this.scheduled = new ArrayList<>();
+
+        tail = empty;
+
+        // clean up to ensure we do not keep references to old tasks
+        cleaner = this.scheduleAtFixedRate(() -> {
+            synchronized (this) {
```
[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java:75]: I use this broad lock here to reduce complexity, i don't think throughput should be an issue here.
So using a finer locking might be possible, but i would like to keep things simple (as long there is no need for it). 
Are you okay with this?

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java (hunk 3 above)
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java:122]: Looks like this is possible, will fix this.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java:48]: will do so

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/ThreadPoolManager.java [hunk 12]
```diff
@@ -115,6 +116,26 @@ protected void modified(Map<String, Object> properties) {
...
+        if (Boolean.getBoolean("org.openhab.core.common.SequentialScheduledExecutorService")) {
```
[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/ThreadPoolManager.java:130]: I did not expect this to get a permanent configuration, so change it to a OSGi service configuration or keep it as system property?

**Comment by joerg1985** [comment]
@kaikreuzer thank you for the review, i addressed most of these points. There are two open points with questions how to continue.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java]: Yes, ok for me as long as throughput is no issue and we do not run into any deadlock situations.

[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/SequentialScheduledExecutorService.java]: As there is only this one object we are synchronizing on and there is no wait for something that happens in another synchronize call to this object, deadlocks should not happen.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/ThreadPoolManager.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/ThreadPoolManager.java:131]: @kaikreuzer sorry for bugging you, but how to continue with this one?

[the remark this answers is not in the document]
Comment (kaikreuzer) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/ThreadPoolManager.java:131]: Sorry - weird, I was sure that I had answered this, but for some reason GitHub does not seem to have stored my comment... So let me try again:

I expect that we will have suc a configuration option not just temporarily, but in the long run. The only thing that might change is the default value of it. For the start, we keep it deactivated, but if we decide one day that it works very nicely and without issues for the majority of users, we might want to activate it by default - while still giving users the chance to manually deactivate it.
Therefore I'd prefer to use our standard configuration mechanism, which is the OSGi service configs.

[the remark this answers is not in the document]
Comment (joerg1985) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/ThreadPoolManager.java]: Okay, i have just updated the PR to use the existing config.

**Comment by joerg1985** [comment]
@kaikreuzer now all the points are addressed, so i think this is ready :rocket:

**Comment by kaikreuzer** [comment]
Thanks @joerg1985!
Do I understand the code correctly, that you didn't introduce any new configuration parameter, but now rather allow using "events" and "rules" to be defined as pools > 0 and this setting makes the new logic become active? If I understand this correctly, this looks like a very neat solution!

**Comment by joerg1985** [comment]
@kaikreuzer yes, this is correct. To enable the new executor the pool size must be set > 0, this allows to enable it only for specific pools.

**Comment by lolodomo** [comment]
> yes, this is correct. To enable the new executor the pool size must be set > 0, this allows to enable it only for specific pools.

Is it documented somewhere?

**Comment by joerg1985** [comment]
@lolodomo no, i am currently testing (and fixing https://github.com/openhab/openhab-core/pull/4247). As soon as there is a stable version, i will write to the Milestone topic to find more testers.

Do you have a concrete place where this should be documented in the future?
